File: tyama_codeeval39.rb
#!/usr/bin/ruby
def seq(n) n.to_s.chars.reduce(0){|s,e|s+=e.to_i**2} end
a=Array.new(200,-1)
a[1]=1
2.step(199){|i|
	if a[i]>-1 then next end
	j=i
	b=[]
	while j!=1 && !b.include?(j)
		b<<j
		j=seq(j)
	end
	b.each{|e|a[e]=(j==1?1:0)}
}
while s=gets do n=s.to_i;n=seq(n) while n>199;p a[n] end

File: tyama_codeeval39.c
a[200];
int seq(n){
	int r=0;
	for(;n;n/=10){
		r+=(n%10)*(n%10);
	}
	return r;
}
int search(n){
	if(n==1)return 2;
	if(a[n])return a[n]==3?1:a[n];
	a[n]=3;
	return a[n]=search(seq(n));
}
main(){
	int i=2;
	for(a[1]=2;i<200;i++)if(!a[i])a[i]=search(i);
	for(;~scanf("%d",&i);printf("%d\n",a[i]-1))for(;i>199;)i=seq(i);
}